test/cpp_api_parity/functional_impl_check.py

```diff
@@ -0,0 +1,239 @@
+
+        cpp_test_fn(arg_dict_file_path, forward_output_file_path)
+        cpp_output = torch.load(forward_output_file_path)
+
+        # Check that forward outputs are equal
+        unit_test_class.assertTrue(
+            torch.allclose(python_output, cpp_output),
+            generate_error_msg("forward output", cpp_output, python_output))
+
+    run_cpp_test_fn_and_check_output()
+
+    # Remove temporary folder that stores C++ outputs
+    assert os.path.exists(cpp_tmp_folder)
+    shutil.rmtree(cpp_tmp_folder)
+
+def compute_functional_name(test_params_dict):
+    def camel_case_to_snake_case(camel_case_str):
+        return re.sub(r'(?<!^)(?=[A-Z])', '_', camel_case_str).lower()
+
+    if 'cpp_options_args' in test_params_dict:
+        # Expected format: `F::FunctionalFuncOptions(...)`
+        return camel_case_to_snake_case(test_params_dict['cpp_options_args'].split('(')[0].replace('F::', '').replace('FuncOptions', ''))
+    elif 'cpp_function_call' in test_params_dict:
+        # Expected format: `F::functional_name(...)`
+        return test_params_dict['cpp_function_call'].split('(')[0].replace('F::', '')
+    else:
+        raise RuntimeError(
+            "`cpp_options_args` or `cpp_function_call` entry must be present in test params dict: {}".format(
+                test_params_dict))
+
+def compute_cpp_function_call(test_params_dict, arg_dict, functional_name):
+    if 'cpp_function_call' in test_params_dict:
+        return test_params_dict['cpp_function_call']
+    elif 'cpp_options_args' in test_params_dict:
+        cpp_forward_args_symbols = [arg_name for arg_name, _ in arg_dict['input'] + arg_dict['target'] + arg_dict['extra_args']]
+        return 'F::{}({}, {})'.format(functional_name, ", ".join(cpp_forward_args_symbols), test_params_dict['cpp_options_args'])
+    else:
+        raise RuntimeError(
+            "`cpp_options_args` or `cpp_function_call` entry must be present in test params dict: {}".format(
+                test_params_dict))
+
+def process_test_params_for_functional(test_params_dict, device, test_instance_class):
+    test_instance = test_instance_class(**test_params_dict)
+    functional_name = compute_functional_name(test_params_dict)
+    assert test_instance.get_name().startswith('test_')
+    functional_variant_name = test_instance.get_name()[5:] + (('_' + device) if device != 'cpu' else '')
+    arg_dict = compute_arg_dict(test_params_dict, test_instance)
+
+    return TorchNNFunctionalTestParams(
+        functional_name=functional_name,
+        functional_variant_name=functional_variant_name,
+        test_instance=test_instance,
+        cpp_function_call=compute_cpp_function_call(test_params_dict, arg_dict, functional_name),
+        arg_dict=arg_dict,
+        has_parity=test_params_dict.get('has_parity', True),
+        device=device,
+        cpp_tmp_folder=tempfile.mkdtemp(),
+    )
+
+torch_nn_test_params_map = {}
```

Comment:
Global state is hard to reason about. Is it necessary for this to be global?

Reply:
One way to avoid this is to make some class object:
```
for test_param in ...:
      FunctionalImplCheck.add_test(blah)
FunctionalImplCheck.build_tests()
FunctionalImplCheck.add_tests_to(TestCppApiParity)
```
This is sort of what the code is doing right now, except the code is modifying the `functional_impl_check` module and essentially treating it like a class with some state (the torch_nn_test_params_map)

Reply:
That would also make some other things more clear, like:
- `functional_impl_check.add_tests` is very different from `utils.add_test`, but they have similar names

Reply:
But I'm not sure if this would actually make the code look nicer. Let me know your thoughts

Reply:
Thanks for looking into it! I moved `torch_nn_test_params_map` into the `unit_test_class` object, so that there is no global state for `functional_impl_check` and `module_impl_check` anymore, but all methods can still access this global map object via `unit_test_class`